src/keystore.rs

```diff
@@ -1,40 +1,44 @@
-use hex::{FromHex, ToHex};
-use serde::{de::Deserializer, ser::Serializer, Deserialize, Serialize};
-use uuid::Uuid;
+use serde::{Deserialize, Deserializer, Serialize};
 
-#[cfg(feature = "geth-compat")]
-use ethereum_types::H160 as Address;
+use crate::{v3, v4};
 
-#[derive(Debug, Deserialize, Serialize)]
-/// This struct represents the deserialized form of an encrypted JSON keystore based on the
-/// [Web3 Secret Storage Definition](https://github.com/ethereum/wiki/wiki/Web3-Secret-Storage-Definition).
-pub struct EthKeystore {
-    #[cfg(feature = "geth-compat")]
-    pub address: Address,
+#[derive(Debug, Serialize)]
+#[serde(tag = "version")]
+pub enum EthKeystore {
+    #[serde(rename = "3")]
+    V3(v3::EthKeystoreV3),
 
-    pub crypto: CryptoJson,
-    pub id: Uuid,
-    pub version: u8,
+    // TODO: add a v4 feature flag
+    #[serde(rename = "4")]
+    V4(v4::EthKeystoreV4),
 }
 
-#[derive(Debug, Deserialize, Serialize)]
-/// Represents the "crypto" part of an encrypted JSON keystore.
-pub struct CryptoJson {
-    pub cipher: String,
-    pub cipherparams: CipherparamsJson,
-    #[serde(serialize_with = "buffer_to_hex", deserialize_with = "hex_to_buffer")]
-    pub ciphertext: Vec<u8>,
-    pub kdf: KdfType,
-    pub kdfparams: KdfparamsType,
-    #[serde(serialize_with = "buffer_to_hex", deserialize_with = "hex_to_buffer")]
-    pub mac: Vec<u8>,
-}
+impl<'de> Deserialize<'de> for EthKeystore {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        let v = serde_json::Value::deserialize(deserializer)?;
 
-#[derive(Debug, Deserialize, Serialize)]
-/// Represents the "cipherparams" part of an encrypted JSON keystore.
-pub struct CipherparamsJson {
-    #[serde(serialize_with = "buffer_to_hex", deserialize_with = "hex_to_buffer")]
-    pub iv: Vec<u8>,
+        match v["version"].as_i64() {
+            // TODO tests/test-key/my_keystore has no version field
+            // should we default to v3 to preserve compatibility, or raise an error?
+            Some(3) | None => {
```

Comment:
this here was one of my main questions. Without `Nnoe`, some of the existing tests fail